composer/utils/remote_uploader.py

```diff
@@ -0,0 +1,172 @@
+
+    @retry(ObjectStoreTransientError, num_attempts=num_attempts)
+    def upload_file(
+        remote_file_name: str,
+        local_file_path: str,
+        overwrite: bool,
+        retry_index: int = 0,
+    ):
+        if retry_index == 0 and not overwrite:
+            try:
+                object_store.get_object_size(remote_file_name)
+            except FileNotFoundError:
+                # Good! It shouldn't exist.
+                pass
+            else:
+                raise FileExistsError(f'Object {remote_file_name} already exists, but overwrite was set to False.')
+        log.info(f'Uploading file {local_file_path} to {remote_file_name}')
+        object_store.upload_object(
+            object_name=remote_file_name,
+            filename=local_file_path,
+        )
+        os.remove(local_file_path)
+
+    def is_parent_process_dead():
+        try:
+            # Sending signal 0 to ppid will raise OSError if ppid is not running
+            # and do nothing otherwise
+            os.kill(parent_process_id, 0)
+        except OSError:
+            return True
+        return False
+
+    while True:
+        try:
+            file_path_to_upload, remote_file_name, overwrite = file_queue.get(block=True, timeout=0.5)
+        except queue.Empty:
+            if is_closed_event.is_set():
+                break
+            elif is_parent_process_dead():
+                log.warning(f'Terminate the uploader worker since parent process died')
+                break
+            else:
+                continue
+
+        # When encountering issues with too much concurrency in uploads, staggering the uploads can help.
+        # This stagger is intended for use when uploading model shards from every rank, and will effectively reduce
+        # the concurrency by a factor of num GPUs per node.
+        local_rank = get_local_rank()
+        local_rank_stagger = int(os.environ.get('COMPOSER_LOCAL_RANK_STAGGER_SECONDS', 0))
+        log.debug(f'Staggering uploads by {local_rank * local_rank_stagger} seconds on {local_rank} local rank.')
+        time.sleep(local_rank * local_rank_stagger)
+
+        upload_file(remote_file_name, file_path_to_upload, overwrite)
+
+
+class RemoteUploader:
+    """Class for uploading a file to object store asynchronously."""
+
+    def __init__(
+        self,
+        remote_folder: str,
+        num_concurrent_uploads: int = 1,
+        num_attempts: int = 3,
+    ):
+        if num_concurrent_uploads < 1:
+            raise ValueError(f'num_concurrent_uploads must be >=1, but got {num_concurrent_uploads}')
+
+        # A folder to use for staging uploads
+        self._tempdir = tempfile.TemporaryDirectory()
+        self._upload_staging_folder = self._tempdir.name
+
+        self.num_attempts = num_attempts
+
+        mp_ctx = multiprocessing.get_context('spawn')
+        manager = mp_ctx.Manager()
+        self.file_queue: queue.Queue = manager.Queue()
```

Comment:
why do we need a filequeue? I think instead of the design where we launch the workers once and then continually poll the file queue, it's even simpler if we just call executor.submit every time we have a new upload request

Reply:
See my another comment. This is because we can't pass object_store to the function that will be executed in a new process. Because you can't pickle an object store. I think that's why the existing remote_uploader_downloader used queues. :)